Docker swarm service integration tests

dariko opened the pull request:
##### SUMMARY
This PR adds integration tests for the `docker_swarm_service` module.
It also fixes a typo on the same module found while writing the tests.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
docker_swarm_service

##### ANSIBLE VERSION
```
ansible 2.8.0.dev0 (docker_swarm_service_integration_test d51d437389) last updated 2018/09/15 14:04:42 (GMT +200)
  config file = None
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/src/ansible/lib/ansible
  executable location = /home/user/src/ansible/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]
```

dariko committed:
integration test for docker_swarm_service

dariko committed:
typo

dariko commented:
Shippable returns this error:
`````
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1537015424.75-158201857908434/AnsiballZ_docker_swarm.py", line 113, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1537015424.75-158201857908434/AnsiballZ_docker_swarm.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1537015424.75-158201857908434/AnsiballZ_docker_swarm.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 525, in <module>
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 520, in main
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 306, in __call__
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 329, in init_swarm
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 389, in __update_swarm
  File "/tmp/ansible_docker_swarm_payload_TDtZ6W/__main__.py", line 375, in __update_spec
KeyError: 'Labels'
`````
on the `[docker_swarm_service : Create a Swarm cluster]` task, which creates
a swarm using the `docker_swarm` module.  

Locally, on Ubuntu 17.10 and Centos7, I run both the `docker_swarm_service`
and the `docker_swarm` tests successfully.

I tried running the tests for `docker_swarm` module on a clean centos7 vm
with the distribution version of docker (`docker-1.13.1-74.git6e3bb8e.el7.centos.x86_64`)
and got the same `KeyError: 'Labels'` error.  
After upgrading docker as per https://docs.docker.com/install/linux/docker-ce/centos/
the was no more errors.

The `docker_swarm` fails when trying to read the `.Spec.Labels` value from a
dict returned from `client.inspect()`, which in turn executes a `GET` to
`/swarm`.  
I checked the responses from the docker apis on both versions and it seems like
the call to `/swarm` on a fresh stack returns:
- no 'Labels' attribute on the older version
  `````
  # curl -s --unix-socket /var/run/docker.sock http:/swarm|jq .Spec
  {
    "Name": "default",
    "Labels": {},
    "Orchestration": {
      "TaskHistoryRetentionLimit": 5
    },
    "Raft": {
      "SnapshotInterval": 10000,
      "KeepOldSnapshots": 0,
      "LogEntriesForSlowFollowers": 500,
      "ElectionTick": 10,
      "HeartbeatTick": 1
    },
    "Dispatcher": {
      "HeartbeatPeriod": 5000000000
    },
    "CAConfig": {
      "NodeCertExpiry": 7776000000000000
    },
    "TaskDefaults": {},
    "EncryptionConfig": {
      "AutoLockManagers": false
    }
  }
  `````
- a 'Label' attribute set to `{}` on the current version
  `````
  # curl -s --unix-socket /var/run/docker.sock http:/swarm|jq .Spec
  {
    "Name": "default",
    "Labels": {},
    "Orchestration": {
      "TaskHistoryRetentionLimit": 5
    },
    "Raft": {
      "SnapshotInterval": 10000,
      "KeepOldSnapshots": 0,
      "LogEntriesForSlowFollowers": 500,
      "ElectionTick": 10,
      "HeartbeatTick": 1
    },
    "Dispatcher": {
      "HeartbeatPeriod": 5000000000
    },
    "CAConfig": {
      "NodeCertExpiry": 7776000000000000
    },
    "TaskDefaults": {},
    "EncryptionConfig": {
      "AutoLockManagers": false
    }
  }
  ````` 

Browsing https://docs.docker.com/engine/api/v1.32/#operation/SwarmInspect
(and `v1.31`) I confirmed that the Labels attribute was added
to the sample response on API version 1.32.

I'd write a patch for the `docker_swarm` module defaulting `client.inspect()['Spec']['Labels']`
to `{}` but shippable [returned no errors](https://app.shippable.com/github/ansible/ansible/runs/78067/summary) when it was last updated.  
Has the shippable environment changed since?
Can someone reproduce this error?  
I feel like I might be missing something...

PS: I noticed that the `setup_docker` fragment installs a recent version of docker but
does not start the service: am I wrong in guessing that it does this to have the
docker client available in the test container while relaying on the parent docker
server (sharing the docker socket?) to execute the subsequent test target tasks?
If so is it possible to control the version of the docker server?

mattclay commented:
@dariko  I've seen the Labels error as well. I don't know why it only shows up on some test runs and not others. The module should be updated to handle the situation where there are no labels.

You are correct that Docker is being installed to have access to the client. The tests are already running in a container so we're relying on the Docker version of the host, which we cannot change during the test run.

mattclay commented on test/integration/targets/docker_swarm_service/tasks/test_swarm_service.yml:
```diff
@@ -0,0 +1,135 @@
+  assert:
+    that:
+    - output is success
+    - output is not changed
```
Make sure to remove the swarm when the test is done:

```yaml
- name: Remove a Swarm cluster
  docker_swarm:
    state: absent
    force: true
```

dariko replied:
added the swarm de-initialization task

mattclay commented:
CI failures are due to two different tracebacks. The first one is this:

```
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1537327837.88-131106585890437/AnsiballZ_docker_swarm_service.py", line 113, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1537327837.88-131106585890437/AnsiballZ_docker_swarm_service.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1537327837.88-131106585890437/AnsiballZ_docker_swarm_service.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_docker_swarm_service_payload_4T6zDE/__main__.py", line 1178, in <module>
  File "/tmp/ansible_docker_swarm_service_payload_4T6zDE/__main__.py", line 1172, in main
  File "/tmp/ansible_docker_swarm_service_payload_4T6zDE/__main__.py", line 1107, in run
  File "/tmp/ansible_docker_swarm_service_payload_4T6zDE/__main__.py", line 1008, in create_service
  File "/usr/local/lib/python2.7/dist-packages/docker/utils/decorators.py", line 34, in wrapper
    return f(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/docker/api/service.py", line 150, in create_service
    rollback_config
  File "/usr/local/lib/python2.7/dist-packages/docker/api/service.py", line 30, in _check_api_features
    raise_version_error('UpdateConfig.order', '1.29')
  File "/usr/local/lib/python2.7/dist-packages/docker/api/service.py", line 11, in raise_version_error
    param, min_version
docker.errors.InvalidVersion: UpdateConfig.order is not supported in API version < 1.29
```

The second occurs when the test is automatically retried:

```
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1537327860.56-258915303905118/AnsiballZ_docker_swarm.py", line 113, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1537327860.56-258915303905118/AnsiballZ_docker_swarm.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1537327860.56-258915303905118/AnsiballZ_docker_swarm.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_docker_swarm_payload_NAzL5T/__main__.py", line 525, in <module>
  File "/tmp/ansible_docker_swarm_payload_NAzL5T/__main__.py", line 520, in main
  File "/tmp/ansible_docker_swarm_payload_NAzL5T/__main__.py", line 306, in __call__
  File "/tmp/ansible_docker_swarm_payload_NAzL5T/__main__.py", line 329, in init_swarm
  File "/tmp/ansible_docker_swarm_payload_NAzL5T/__main__.py", line 389, in __update_swarm
  File "/tmp/ansible_docker_swarm_payload_NAzL5T/__main__.py", line 375, in __update_spec
KeyError: 'Labels'
```

dariko committed:
ensure stack de-initialization

dariko commented:
@mattclay thank you, I've committed the changes you suggested.
I didn't notice the `UpdateConfig.Order` error (I was only looking at the one reported on the console by shippable): I might be able to fix it using a version check for all the `update_*` parameters.

dariko committed:
Set default value for 'configs' parameter to None

Docker-py uses None as a default value for configs.
Using the same default here allows to create services on older docker
setups (docker_api<1.30).

dariko committed:
Set default value for 'update_order' parameter to None

Docker-py uses None as a default value for update_order.
Using the same default here allows to create services on older docker
setups (docker_api<1.29)

dariko committed:
Set default value for 'publish.mode' parameter to None

Docker-py uses None as a default value for publish_mode.
Using the same default here allows to create services on older docker
setups (docker_api<1.32)

dariko committed:
Allow tests to run on older version of docker.

dariko commented:
@mattclay

The `config`, `update_order` and `publish.mode` parameter default values different from their docker-py counterparts and invalid for older docker apis:
I changed their default to `None`.

After that I got into the same problem you encountere on #45807.  
Diving deeper into the [context deadline exceeded](https://github.com/moby/moby/search?q=context+deadline+exceeded&type=Issues) I have not found a definitive answer, but good old trial/error on a box with docker 1.31 had me notice how the `docker_swarm` cluster leave request did in fact get enqueued by the docker daemon, it just was executed with a non-insignificant delay.  

The tests run successfully, leaving the parent in clean state, retrying the `docker_swarm` invocation 3 times (some docker timeout gets cleared, I think).  
This is definitely not ideal, and I don't know if it is acceptable, but it works both on my test and in shippable.

It would be nice to be able to run the tests using a more recent version of docker: while testing on older versions is useful, testing on the last stable one is what I'd expect.  
Unfortunately a cursory look at shippable docs seems to show only support for test running in docker.

dariko commented:
[Here](https://github.com/moby/moby/issues/35337#issuecomment-341413173) I found a nice explanation of what might be going wrong, suggesting two workarounds:
> If you're in this situation and want to "destroy" the swarm, you can try stopping the daemon, and
> removing the /var/lib/docker/swarm/ directory. You can also re-initialize the swarm (which may 
> preserve the information that was present in the swarm), using docker swarm init --force-new-cluster 
> on the manager node.

The first one would definitely work, but AFAIK can't be used on ansible's test infrastructure, as it would require direct access to the node running the test container.

Following the `init --force-new-cluster` path:
`````
- name: Force create new swarm
  docker_swarm:
    state: present
    force: true
    advertise_addr: "{{ansible_default_ipv4.address}}"
- name: Remove a Swarm cluster
  docker_swarm:
     state: absent
     force: true
`````
returns the same [context deadline exceeded](https://app.shippable.com/github/ansible/ansible/runs/86201/52/tests) error.

Maybe `docker_swarm` module should manage this corner case?
In the meantime I'll leave this PR playbooks in the _ugly but working_ state.

dariko committed:
remove workarounds for old docker versions

dariko committed:
test correct swarm cleanup

dariko commented:
@mattclay
Have the docker version changed on the shippable nodes?
Now this test runs without the workarounds introduced for the older docker
version.
I added a test to ensure the docker daemon is left in a clean state.

PS: I don't know how/if to report this, but I got a strange error on the first steps
of a CI run on freebsd 11.1:
https://app.shippable.com/github/ansible/ansible/runs/88396/45/console
`````
Generating Job
Provision job node
Setup
Job node info
 uname -a
 uptime
 df -h
 free -m
 echo 'Docker daemon successfully running on the host'
 sudo docker info
  Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
 Cleanup 
`````

but at the next push the test run correctly.

mattclay commented:
@dariko Yes, we were finally able to upgrade our Shippable nodes to the latest version, so we're now using a much newer version of Docker.

We don't have Docker running on our FreeBSD, RHEL or MacOS test systems -- they're actual VMs instead of Docker containers on Shippable.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_service.py:
```diff
@@ -177,7 +177,7 @@
     - Every item must be a dictionary exposing the keys config_id, config_name, filename, uid (defaults to 0), gid (defaults to 0), mode (defaults to 0o444)
     - Maps docker service --config option.
-    default: []
+    default: null
```
How about just leaving it away (and removing `default=None` in args_spec)? The default value is `None` anyway if the argument is not required.

dariko replied:
@felixfontein  Thank you, I didn't know that; I'm gonna change them

felixfontein reviewed (commented):
Out of curiosity: why do you change the behavior w.r.t. some of the arguments in this PR? I guess this doesn't change the module's behavior (if yes, some changelog entry is needed!), but I don't know the code well enough to be sure :)

Are some of the changes included here bugfixes (which deserve to be in their own PR so they can be backported)?

dariko commented:
@felixfontein  about the changed defaults
There were problems with some older docker API version which did not support those parameters for which I changed the default. Pointing the module to those API versions caused the module to fail as the request still included the relative section (even if empty) and the daemon threw an error (50x, I think).
I don't remember if this was the case for all the parameter which I changed, or if I decided to remodel other parameter as to map the defaults to the API response for a GET to a "clean" service.

For example the parameter `update_order`, which maps to the API's `UpdateConfig.Order`, was introduced in [API 1.29](https://github.com/moby/moby/pull/30261/files#diff-728a7fe135201867a6b4b645adc9553dR538) (docker 17.05). Now the module delegates the default of `update_order` to docker, which is `stop-first` as was hardcoded in here, but it can also be executed pointing to a daemon older than 17.05.
The same is true for `configs` (introduced in 1.30) and `port.mode` (since 1.25).

I think there is no behaviour change, as in the interpreted meaning of the defaults is the same, it is now possible to use the module event with older version of docker.
I found out this problems only because when I began writing this PR the images used on shippable included an old docker version (I think something <=1.13), so these are bugfixes of bugs which no one using a reasonable setup will ever have found.

dariko commented:
@felixfontein I'm definitely familiar with the docker-py/docker python libraries version problems, and I don't think those apply here.

I've retraced the steps which had me proposed the parameter default changes.

(here)[https://gist.github.com/dariko/3fdcc08227993ccee23d150c8f915f39] is a log of the installation/execution of the module on a fresh centos7-cloud with ansible 2.7.0, latest docker python module and distribution provided docker (1.13). Called with minimal parameters the module fails with this error

`````
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1540053836.44-186112295070907/AnsiballZ_docker_swarm_service.py", line 113, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1540053836.44-186112295070907/AnsiballZ_docker_swarm_service.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1540053836.44-186112295070907/AnsiballZ_docker_swarm_service.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_docker_swarm_service_payload_fLUl1n/__main__.py", line 1178, in <module>
  File "/tmp/ansible_docker_swarm_service_payload_fLUl1n/__main__.py", line 1172, in main
  File "/tmp/ansible_docker_swarm_service_payload_fLUl1n/__main__.py", line 1107, in run
  File "/tmp/ansible_docker_swarm_service_payload_fLUl1n/__main__.py", line 1008, in create_service
  File "/usr/lib/python2.7/site-packages/docker/utils/decorators.py", line 34, in wrapper
    return f(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/docker/api/service.py", line 150, in create_service
    rollback_config
  File "/usr/lib/python2.7/site-packages/docker/api/service.py", line 30, in _check_api_features
    raise_version_error('UpdateConfig.order', '1.29')
  File "/usr/lib/python2.7/site-packages/docker/api/service.py", line 11, in raise_version_error
    param, min_version
docker.errors.InvalidVersion: UpdateConfig.order is not supported in API version < 1.29
`````

Manully updating the installed module to implement 63641a91e18d3aa0d7bc92af69e885b18c75c5f1 yield this other error:
`````
[Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1540054391.08-201999029627167/AnsiballZ_docker_swarm_service.py", line 113, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1540054391.08-201999029627167/AnsiballZ_docker_swarm_service.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1540054391.08-201999029627167/AnsiballZ_docker_swarm_service.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_docker_swarm_service_payload_J7eJVJ/__main__.py", line 1178, in <module>
  File "/tmp/ansible_docker_swarm_service_payload_J7eJVJ/__main__.py", line 1172, in main
  File "/tmp/ansible_docker_swarm_service_payload_J7eJVJ/__main__.py", line 1107, in run
  File "/tmp/ansible_docker_swarm_service_payload_J7eJVJ/__main__.py", line 1008, in create_service
  File "/usr/lib/python2.7/site-packages/docker/utils/decorators.py", line 34, in wrapper
    return f(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/docker/api/service.py", line 150, in create_service
    rollback_config
  File "/usr/lib/python2.7/site-packages/docker/api/service.py", line 83, in _check_api_features
    raise_version_error('ContainerSpec.configs', '1.30')
  File "/usr/lib/python2.7/site-packages/docker/api/service.py", line 11, in raise_version_error
    param, min_version
docker.errors.InvalidVersion: ContainerSpec.configs is not supported in API version < 1.30
`````

In the module itself there is a (validation function)[https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_swarm_service.py#L1014-L1037] which has the module fail gracefully if some parameters unsupported by the targeted docker daemon are called. 
That function already contains a check on the `config` parameter, allowing it to be specified only if API>=1.30. Still, *the module fails because it wrongly assumes the server would be willing to accept an empty 'configs' parameter, which is not the case for docker<=1.30*.
The same reasoning can be applied for the `update_order`/`UpdateConfig.Order` error.

Still, I don't know about the ansible management processes and I don't know if this requires a changelog entry; I'm pretty sure these commits should be backported, I have no idea how to do it and I still haven't looked into it.

In all this I'm trying to keep in mind that either no one tried using this module in such an environment (which gives me back a tiny bit of faith in humanity :smile:) or no one reported it (.. or I haven't noticed so), but as the linked log shows the module is currently simply broken for older docker versions.

felixfontein commented:
All PRs should have a changelog, see [here](https://groups.google.com/forum/#!msg/ansible-devel/uYuIADYa5vs/Cby5dnDRAAAJ;context-place=forum/ansible-devel). For something that is completely not user-facing (test changes, internal module changes which are not visible) I guess it can be left away, but for a bugfix like this one it should be included :)

Anyway, the process is that bugfixes are usually backported (in this case to 2.7 so that it will be contained in 2.7.1).

And that nobody yet complained about it might also be because 2.7 was just released a couple of weeks ago, so not too many had the chance to try it out :) 2.7.1 is probably released next week (if I didn't count wrong), so getting an improved version of the module into it is a good idea!

dariko committed:
changelog fragment for docker_swarm_service defaults change

dariko commented:
Unluckily I only now managed to find some time. I added a changelog fragment for the changed parameters default.

felixfontein reviewed (approved):
No problem :)

shipit

gundalow commented:
@dariko Thank you for the PR
@felixfontein Thank you, as always, for the reviews.

As this is a feature merged into `devel` for release in Ansible 2.8
